csv.go

```diff
@@ -0,0 +1,157 @@
+package ftdc
+
+import (
+	"context"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/mongodb/mongo-go-driver/bson"
+	"github.com/pkg/errors"
+)
+
+func (c *Chunk) getFieldNames() []string {
+	fieldNames := make([]string, len(c.metrics))
+	for idx, m := range c.metrics {
+		fieldNames[idx] = m.Key()
+	}
+	return fieldNames
+}
+
+func (c *Chunk) getRecord(i int) []string {
+	fields := make([]string, len(c.metrics))
+	for idx, m := range c.metrics {
+		switch m.originalType {
+		case bson.TypeDouble, bson.TypeInt32, bson.TypeInt64, bson.TypeBoolean, bson.TypeTimestamp:
+			fields[idx] = strconv.FormatInt(m.Values[i], 10)
+		case bson.TypeDateTime:
+			fields[idx] = time.Unix(m.Values[i]/1000, 0).Format(time.RFC3339)
+		}
+	}
+	return fields
+}
+
+// WriteCSV exports the contents of a stream of chunks as CSV. Returns
+// an error if the number of metrics changes between points, or if
+// there are any errors writing data.
+func WriteCSV(ctx context.Context, iter *ChunkIterator, writer io.Writer) error {
+	var numFields int
+	csvw := csv.NewWriter(writer)
+	for iter.Next(ctx) {
+		chunk := iter.Chunk()
+		if numFields == 0 {
+			fieldNames := chunk.getFieldNames()
+			if err := csvw.Write(fieldNames); err != nil {
+				return errors.Wrap(err, "problem writing field names")
+			}
+			numFields = len(fieldNames)
+		} else if numFields != len(chunk.metrics) {
+			return errors.New("unexpected schema change detected")
+		}
+
+		for i := 0; i < chunk.nPoints; i++ {
+			record := chunk.getRecord(i)
+			if err := csvw.Write(record); err != nil {
+				return errors.Wrapf(err, "problem writing csv record %d of %d", i, chunk.nPoints)
+			}
+		}
+		csvw.Flush()
+		if err := csvw.Error(); err != nil {
+			return errors.Wrapf(err, "problem flushing csv data")
+		}
+	}
+	if err := iter.Err(); err != nil {
+		return errors.Wrap(err, "problem reading chunks")
+	}
+
+	return nil
+}
+
+func getCSVFile(prefix string, count int) (io.WriteCloser, error) {
+	fn := fmt.Sprintf("%s.%d.csv", prefix, count)
+	writer, err := os.Create(fn)
+	if err != nil {
+		return nil, errors.Wrapf(err, "provlem opening file %s", fn)
+	}
+	return writer, nil
+}
+
+// DumpCSV writes a sequence of chunks to CSV files, creating new
+// files if the iterator detects a schema change, using only the
+// number of fields in the chunk to detect schema changes. DumpCSV
+// writes a header row to each file.
+//
+// The file names are constructed as "prefix.<count>.csv".
+func DumpCSV(ctx context.Context, iter *ChunkIterator, prefix string) error {
```

Comment:
maybe call this `DynamicCSV` to be consistent with `DynamicCollector`?

Reply:
I kind of like it as is. Its less about dynamic/not and more about the chunking.

maybe it makes sense to add a max number of rows per file as well to make it less about schema change.

Reply:
Well it only creates a new csv file when the schema changes, so as is it is all about schema change. 

If we do add max number of rows, we would still have to change when the schema changes, correct?

Reply:
yep.